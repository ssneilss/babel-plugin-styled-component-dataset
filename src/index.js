import { basename, extname, dirname } from 'path';
import * as t from 'babel-types';

const importLocalName = (name, state) => {
  let localName = name === 'default' ? 'styled' : name;

  state.file.path.traverse({
    ImportDeclaration: {
      exit(path) {
        const { node } = path;

        if (node.source.value === 'styled-components') {
          path.get('specifiers').forEach((specifier) => {
            if (specifier.isImportDefaultSpecifier()) {
              localName = specifier.node.local.name;
            }

            if (specifier.isImportSpecifier() && specifier.node.imported.name === name) {
              localName = specifier.node.local.name;
            }

            if (specifier.isImportNamespaceSpecifier()) {
              localName = specifier.node.local.name;
            }
          });
        }
      },
    },
  });

  return localName;
};

const isStyled = (tag, state) => {
  if (
    t.isCallExpression(tag) &&
    t.isMemberExpression(tag.callee) &&
    tag.callee.property.name !== 'default' /** ignore default for #93 below */
  ) {
    // styled.something()
    return isStyled(tag.callee.object, state);
  }

  return (
    (t.isMemberExpression(tag) && tag.object.name === importLocalName('default', state)) ||
    (t.isCallExpression(tag) && tag.callee.name === importLocalName('default', state)) ||
    (t.isCallExpression(tag) &&
      t.isCallExpression(tag.callee.object) &&
      tag.callee.object.callee.name === 'require' &&
      tag.callee.object.arguments[0].value === 'styled-components')
  );
};

const getPrefix = (state) => {
  let prefix;
  prefix = basename(state.file.opts.filename, extname(state.file.opts.filename));
  if (prefix === 'index') {
    [prefix] = dirname(state.file.opts.filename)
      .split('/')
      .reverse();
  }
  return prefix;
};

const getName = (state, path) => {
  let namedNode;
  path.find((p) => {
    // const X = styled
    if (p.isAssignmentExpression()) {
      namedNode = p.node.left;
      // const X = { Y: styled }
    } else if (p.isObjectProperty()) {
      namedNode = p.node.key;
      // class Y { (static) X = styled }
    } else if (p.isClassProperty()) {
      namedNode = p.node.key;
      // let X; X = styled
    } else if (p.isVariableDeclarator()) {
      namedNode = p.node.id;
    } else if (p.isStatement()) {
      // we've hit a statement, we should stop crawling up
      return true;
    }

    if (namedNode) return true;
    return false;
  });

  // foo.bar -> bar
  if (t.isMemberExpression(namedNode)) {
    namedNode = namedNode.property;
  }

  return t.isIdentifier(namedNode) ? namedNode.name : undefined;
};

export default function () {
  return {
    visitor: {
      TaggedTemplateExpression(path, state) {
        if (!isStyled(path.node.tag, state)) return;

        const key = state.opts.key ? `'${state.opts.key}'` : "'data-id'";
        const prefix = getPrefix(state);
        const name = getName(state, path);

        // eslint-disable-next-line no-param-reassign
        path.node.tag = t.callExpression(t.memberExpression(path.node.tag, t.identifier('attrs')), [
          t.objectExpression([
            t.objectProperty(t.identifier(key), t.stringLiteral(`${prefix}_${name}`)),
          ]),
        ]);
      },
    },
  };
}
