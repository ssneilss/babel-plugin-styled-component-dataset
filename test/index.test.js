/* eslint-env jest */
import fs from 'fs';
import path from 'path';
import { transformFileSync } from 'babel-core';
import '../src';

describe('should transform styled-component', () => {
  const fixturesDir = path.join(__dirname, 'fixtures');
  fs.readdirSync(fixturesDir).forEach((caseName) => {
    it(`should ${caseName.replace(/-/g, ' ')}`, () => {
      const { code } = transformFileSync(`${path.join(fixturesDir, caseName)}/index.js`);
      expect(code).toMatchSnapshot();
    });
  });
});
