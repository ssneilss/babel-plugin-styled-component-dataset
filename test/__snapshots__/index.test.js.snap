// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should transform styled-component should add custom dataset to styled component 1`] = `
"import s from 'styled-components';

const Container1 = s.div.attrs({
  data-test-key: 'add-custom-dataset-to-styled-component_Container1'
})\`\`;"
`;

exports[`should transform styled-component should add dataset to hoisted styled component 1`] = `
"import s from 'styled-components';

const Container1 = s.div.attrs({
  data-id: 'add-dataset-to-hoisted-styled-component_Container1'
})\`\`;

const Container2 = true ? s.div.attrs({
  data-id: 'add-dataset-to-hoisted-styled-component_Container2'
})\`\` : s.div.attrs({
  data-id: 'add-dataset-to-hoisted-styled-component_Container2'
})\`\`;

const components = {
  Container1,
  Container2
};"
`;

exports[`should transform styled-component should add dataset to imported styled component 1`] = `
"const Container1 = s.div.attrs({
  data-id: 'add-dataset-to-imported-styled-component_Container1'
})\`\`;

import { default as s, css } from 'styled-components';"
`;

exports[`should transform styled-component should add dataset to styled component 1`] = `
"import styled from 'styled-components';

const Container1 = styled.div.attrs({
  data-id: 'add-dataset-to-styled-component_Container1'
})\`\`;

const Container2 = styled('div').attrs({
  data-id: 'add-dataset-to-styled-component_Container2'
})\`\`;

const Container3 = require('styled-components').default('div').attrs({
  data-id: 'add-dataset-to-styled-component_Container3'
})\`\`;"
`;

exports[`should transform styled-component should add dataset to styled component with attrs 1`] = `
"import styled from 'styled-components';

const Container1 = styled.div.attrs({ 'data-test': 'test' }).attrs({
  data-id: 'add-dataset-to-styled-component-with-attrs_Container1'
})\`\`;

const Container2 = styled.div.withConfig({ prop: 'value' }).attrs({ 'data-test': 'test' }).attrs({
  data-id: 'add-dataset-to-styled-component-with-attrs_Container2'
})\`\`;

const Container4 = styled('div').attrs({ 'data-test': 'test' }).attrs({
  data-id: 'add-dataset-to-styled-component-with-attrs_Container4'
})\`\`;

const Container3 = styled(Container).attrs({ 'data-test': 'test' }).attrs({
  data-id: 'add-dataset-to-styled-component-with-attrs_Container3'
})\`\`;"
`;

exports[`should transform styled-component should ignore adding dataset 1`] = `
"import styled, { css, injectGlobal, isStyledComponent, keyframes } from 'styled-components';

const Container = styled.div.attrs({
  data-test-key: 'ignore-adding-dataset_Container'
})\`\`;

const style = css\`\`;

const keyframe = keyframes\`\`;

injectGlobal\`\`;

isStyledComponent(Container);

const Container1 = notStyled.div\`\`;"
`;
